### PR TITLE
Add multiple etcd support.

### LINKED ISSUE #1231: [discuss] feature: support etcd cluster in Apache APISIX.
The etcd cluster is highly available, so Apache APISIX should allow multiple etcd addresses to be configured to connect to the etcd cluster.

How about this new style?

```yaml
etcd:
  host: 
      - "http://127.0.0.1:2379"   # multiple etcd address
      - "http://127.0.0.1:2380"  
  prefix: "/apisix"               # apisix configurations prefix
  timeout: 3                      # 3 seconds
```

BTW, the driver `resty-etcd` has supported this feature.

https://github.com/iresty/lua-resty-etcd/blob/master/t/v2/cluster.t#L52

### PR DESCRIPTION
**Pull request by Akayeshmantha**
Add multiple etcd address support

### Issues resolved

Fix #1231

### CODE REVIEW

**Comment by Akayeshmantha** [comment]
Hi @membphis 

An issue in travis again hm 

`An error occurred while generating the build script.`

Any idea ?

**Comment by membphis** [comment]
> An error occurred while generating the build script.

I have restarted the Travis job. wait for a moment

File: conf/config.yaml [hunk 1]
```diff
@@ -113,7 +113,8 @@ nginx_config:                     # config for render the template to genarate n
       - 'unix:'
 
 etcd:
-  host: "http://127.0.0.1:2379"   # etcd address
+  host:
+    - "http://127.0.0.1:2379"     # etcd address
```
Comment (membphis) [inline conf/config.yaml]: Add a comment, here is allowing multiple etcd addresses

**Comment by Akayeshmantha** [comment]
@membphis  am so sorry for the PR rebase problem but the rebase was not done correctly for the first time it created a little mess there so I closed that PR.

File: bin/apisix [hunk 2]
```diff
@@ -706,28 +706,33 @@ local function init_etcd(show_output)
...
 
-    for _, dir_name in ipairs({"/routes", "/upstreams", "/services",
-                               "/plugins", "/consumers", "/node_status",
-                               "/ssl", "/global_rules", "/stream_routes",
-                               "/proto"}) do
-        local cmd = "curl " .. uri .. dir_name
+    for _, host in ipairs(yaml_conf.etcd.host) do
+        uri = host .. "/v2/keys" .. (etcd_conf.prefix or "")
+
+        for _, dir_name in ipairs({"/routes", "/upstreams", "/services",
```
Comment (membphis) [inline bin/apisix]: The current process is wrong.

That's fake code, please take a look.

```lua
-- fake code
for i, host in ipairs(yaml_conf.etcd.host) do
    local is_success = true
    for _, dir_name in ipairs({"/routes", "/upstreams", ...}) do
        local ok = ...
        if not ok then
            is_success = false
            break
        end
    end

    if is_success then
        break
    end
end
```

Comment (Akayeshmantha) [reply bin/apisix]: @membphis fixed this

File: bin/apisix [hunk 3]
```diff
@@ -706,28 +706,40 @@ local function init_etcd(show_output)
...
 
-        local res = exec(cmd)
-        if not res:find("index", 1, true)
-           and not res:find("createdIndex", 1, true) then
-            error(cmd .. "\n" .. res)
+            local res = exec(cmd)
+            if not res:find("index", 1, true)
+                    and not res:find("createdIndex", 1, true) then
+                is_success = false
+                error(cmd .. "\n" .. res)
```
Comment (membphis) [inline bin/apisix]: If running the command fails, we should just print the warning message here.

Using other etcd instances may succeed.

When it finally failed, we will throw an error message.

Comment (Akayeshmantha) [reply bin/apisix]: @membphis  oh okays I thought it was only an error log print.

File: bin/apisix [hunk 4]
```diff
@@ -706,28 +706,43 @@ local function init_etcd(show_output)
     end
 
     local etcd_conf = yaml_conf.etcd
-    local uri = etcd_conf.host .. "/v2/keys" .. (etcd_conf.prefix or "")
 
     local timeout = etcd_conf.timeout or 3
+    local uri
+    local hostCount = table.getn(yaml_conf.etcd.host)
```
Comment (membphis) [inline bin/apisix]: We should always use `#` to get the length of the array object.

Comment (membphis) [inline bin/apisix]: name style: `hostCount ` -> `host_count`

File: bin/apisix [hunk 5]
```diff
@@ -706,28 +706,43 @@ local function init_etcd(show_output)
...
                     .. " --max-time " .. timeout * 2 .. " --retry 1 2>&1"
 
-        local res = exec(cmd)
-        if not res:find("index", 1, true)
-           and not res:find("createdIndex", 1, true) then
-            error(cmd .. "\n" .. res)
+            local res = exec(cmd)
+            if not res:find("index", 1, true)
+                    and not res:find("createdIndex", 1, true) then
+                is_success = false;
```
Comment (membphis) [inline bin/apisix]: code style: no separator is used by default.

**Comment by Akayeshmantha** [comment]
@membphis  done

File: bin/apisix [hunk 6]
```diff
@@ -706,28 +706,43 @@ local function init_etcd(show_output)
     end
 
     local etcd_conf = yaml_conf.etcd
-    local uri = etcd_conf.host .. "/v2/keys" .. (etcd_conf.prefix or "")
 
     local timeout = etcd_conf.timeout or 3
+    local uri
+    local host_count = #(yaml_conf.etcd.host)
```
Comment (membphis) [inline bin/apisix]: Apache APISIX users may use the old `conf/config.yaml` and I think we should check the data type first.

For example Lua code:

```lua
if type(yaml_conf.etcd.host) == "string" then
    yaml_conf.etcd.host = {yaml_conf.etcd.host}
end

local host_count = #(yaml_conf.etcd.host)
...
```

Comment (Akayeshmantha) [reply bin/apisix]: Yes that will be good

Comment (Akayeshmantha) [reply bin/apisix]: @membphis  done

File: bin/apisix [hunk 7]
```diff
@@ -706,28 +706,48 @@ local function init_etcd(show_output)
...
-    for _, dir_name in ipairs({"/routes", "/upstreams", "/services",
-                               "/plugins", "/consumers", "/node_status",
-                               "/ssl", "/global_rules", "/stream_routes",
-                               "/proto"}) do
-        local cmd = "curl " .. uri .. dir_name
+    --convert old single etcd config to multiple etcd config
+    if type(yaml_conf.etcd.host) == "string" then
+        yaml_conf.etcd.host = {yaml_conf.etcd.host}
+    end
+
```
Comment (membphis) [inline bin/apisix]: move https://github.com/apache/incubator-apisix/pull/1283/files#diff-62edadffc237f13dc28a694080d293a8R712 to here

File: bin/apisix [hunk 8]
```diff
@@ -706,28 +706,48 @@ local function init_etcd(show_output)
     end
 
     local etcd_conf = yaml_conf.etcd
-    local uri = etcd_conf.host .. "/v2/keys" .. (etcd_conf.prefix or "")
 
     local timeout = etcd_conf.timeout or 3
+    local uri
+    local host_count = #(yaml_conf.etcd.host)
```
Comment (membphis) [inline bin/apisix]: if the `yaml_conf.etcd.host` is a string object, it will get the string length of it.

Comment (Akayeshmantha) [reply bin/apisix]: oh sure @membphis

**Comment by Akayeshmantha** [comment]
thanks @membphis . It was a great learning experience

**Comment by Akayeshmantha** [comment]
@membphis  any documents to be updated ?

**Comment by moonming** [comment]
> @membphis  any documents to be updated ?

I think you can add comments and a example in config.yaml

**Comment by Akayeshmantha** [comment]
added more comments

**Comment by membphis** [comment]
@Akayeshmantha merged, many thx

**Pull request merged**